pkg/controllers/deprovisioning/orchestration/queue.go

```diff
@@ -0,0 +1,318 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package orchestration
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/avast/retry-go"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/logging"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	deprovisioningevents "github.com/aws/karpenter-core/pkg/controllers/deprovisioning/events"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/utils/nodeclaim"
+)
+
+const (
+	queueBaseDelay   = 1 * time.Second
+	queueMaxDelay    = 10 * time.Second
+	maxRetryDuration = 10 * time.Minute
+)
+
+type Command struct {
+	ReplacementKeys []*NodeClaimKey
+	Candidates      []*state.StateNode
+	Reason          string    // reason is used for metrics
+	TimeAdded       time.Time // timeAdded is used to track timeouts
+	LastError       error
+}
+
+// NodeClaimKey wraps a nodeclaim.Key with an initialized field to save on readiness checks and identify
+// when a nodeclaim is first initialized for metrics and events.
+type NodeClaimKey struct {
+	nodeclaim.Key
+	// Use a bool track if a node has already been initialized. This intentionally does not capture nodes that go
+	// initialized then go NotReady after as other pods can schedule to this node as well.
+	Initialized bool
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	// providerID -> command, maps a candidate to its command. Each command has a list of candidates that can be used
+	// to map to itself.
+	candidateProviderIDToCommand map[string]*Command
+
+	kubeClient client.Client
+	recorder   events.Recorder
+	cluster    *state.Cluster
+	clock      clock.Clock
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate deprovisioning commands
+func NewQueue(ctx context.Context, kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock, testingMode bool) *Queue {
+	queue := &Queue{
+		RateLimitingInterface:        workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		candidateProviderIDToCommand: map[string]*Command{},
+		kubeClient:                   kubeClient,
+		recorder:                     recorder,
+		cluster:                      cluster,
+		clock:                        clock,
+	}
+	// If testing, we don't actually need to start the queue. We can call each function individually instead of relying
+	// on the rate limiting interface itself.
+	if !testingMode {
+		go queue.Start(logging.WithLogger(ctx, logging.FromContext(ctx).Named("deprovisioning")))
+	}
```

Comment:
I would avoid implementing a testing mode at all costs. Perhaps you can just make the struct members public, and test it that way? Or you could test using the same package if you wanted to keep things private.

Reply:
Yeah, I kept all struct members public, but I wanted to make sure that I could test the deprovisioning controller without having the queue run async, since it's harder to control the tests that way. We currently do a `simulationMode` with the Provisioner.Scheduler, which I was kind of basing this off of. What's the problem with a testing mode?